(MODULES-9428) Clarify composite namevar edgecases and add examples (#140)
puppetlabs/puppet-specifications: language/resource-api/README.md

* `autorequire`, `autobefore`, `autosubscribe`, and `autonotify`: a hash mapping resource types to titles. The titles must either be constants, or, if the value starts with a dollar sign, a reference to the value of an attribute. If the specified resources exist in the catalog, Puppet will create the relationsships requested here.
* `features`: a list of API feature names, specifying which optional parts of this spec the provider supports. Currently defined features: `canonicalize`, `simple_get_filter`, and `supports_noop`. See below for details.

For autoloading to work, this code needs to go into `lib/puppet/type/<name>.rb` in your module.

###  Composite Namevars ("title_patterns")

Each resource being managed must be identified by a unique title. Usually this is fairly straightforward and a single attribute can be used to act as an identifier. Sometimes though, you need a composite of two or more attributes to uniquely identify the resource you want to manage.

> If multiple attributes are defined with the `namevar` behaviour, the type SHOULD specify `title_patterns` that will tell Resource API how to get at the attributes from the title. If `title_patterns` is not specified a default pattern is applied, and matches the entire title against the first declared `namevar`.

Reviewer: Is this `SHOULD` really a `MUST` - if not, how would it otherwise deal with it?

Author: see below/

> > Note: Not supplying a title_pattern that assigns values to all namevars will severely restrict the usability of the provider.

Reviewer: So what should you instead do?

Author: As the text says: if `title_patterns` is not specified a default pattern is applied, and matches the entire title against the first declared `namevar`. This will mean that in puppet code all other namevars need to be specified as attributes in the body of the resource.

Reviewer: I still feel line 80 is dangling - comes across as a threat as I don't really understand what it is telling me. I understand line 79, it is telling me what I should/must do.

Reviewer: Ah... I think it is saying "A supplied title pattern MUST assign values to all namevars, as not doing this will severely restrict the usability of the provider."
Did I get it?

Reviewer: Sorry to harp on this, but since I don't know how it actually works and would like to know... I keep coming back to this :-)
When you say "As the text says: if title_patterns is not specified a default pattern is applied, and matches the entire title against the first declared namevar. This will mean that in puppet code all other namevars need to be specified as attributes in the body of the resource." - is that in any way recommended way of doing it? Are people doing it? If so, you may want to include something about this - "In legacy code this may have been implemented as <some variant of what you wrote> ... This is discouraged (deprecated even?) as it strongly restricts the usability of the provider"

Author: I'm trying to toe the line between "it's not required" and "if you try and it breaks, you may keep both pieces".

I should just stamp down on it and make not having a `title_pattern` that assigns all namevars an error.

Reviewer: Yes, and then change the `SHOULD` to `MUST` - we take no responsibility, nor do we tell you what will happen - it is simply undefined and W.R.O.N.G.

Author: Yeah, I'll do that - even if it's possibly not something we can actively check on the backend. :man_shrugging:

}
```

> The provider should return a string containing all namevar values in a string matching one of the title patterns. This allows the runtime environment to present resources in a way familiar to the user.

Reviewer 2: ```suggestion
The provider returns a string containing all namevar values in a string matching one of the title patterns.
```

Reviewer 2: I don't think the last line adds anything, and avoid saying "the user", as they are the ones reading it.

Reviewer: Actually, now that this sentence was called out and I read it again, it does not really make sense - "returning a string containing all namevar values in a string" ... what does that actually mean. And the original said "should return", as in a requirement on the author of a provider; unclear if this is something an implementor of a resource type can do anything about...

Can meaning be made a more clear?

Reviewer 3: > I don't think the last line adds anything, and avoid saying "the user", as they are the ones reading it.

For clarity, this documentation is directed at a module author, they must ensure that the provider that they are writing returns a `namevar` string that matches a `title-pattern` in a way that  "the user" (the eventual end user of the new module) understands.

Author: I've rewritten this section extensively

> The provider should return a string containing all namevar values in a string matching one of the title patterns. This allows the runtime environment to present resources in a way familiar to the user.

Reviewer: ```suggestion
The provider should return a single formatted string representing all namevar values in such a way that it matches one of the title patterns. This allows the runtime environment to present resources in a way familiar to the user.
```
Now I get it - how about this suggestion?

Author: I've added the where and how bit, too.

* If the type has more than one namevar, each resource hash must have a `:title` key with a single formatted string representing all namevar values. The title must match one of the title patterns. The namevar values computed from that pattern must match their counterparts in the hash.
  > Note: This value allows the runtime environment to present resources in a way familiar to the user.

  > Example: Referring back to the `software` example above, resources returned from the `get` method require a `title:` value matching the `package:` and `manager:` values:
  > ```
  > {
  >     title: 'php-yum',
  >     package: 'php',
  >     manager: 'yum',
  >     ensure: 'present',
  > }
  > ```

The `set` method updates resources to a new state. The `changes` parameter gets passed a hash of change requests, keyed by the resource's name. Each value is another hash with the optional `:is` and `:should` keys. At least one of the two has to be specified. The values will be of the same shape as those returned by `get`. After the `set`, all resources should be in the state defined by the `:should` values.

> Example for a `changes` request:
> ```
> changes = {

> >   'name' => {

Reviewer 3: I think this example would be clearer if the name value was something like 'foo' . 
```
changes = {
  'foo' => {
    is: {
      name: 'foo',
      ensure: 'present',
      value: 'original value',
    },
    should: {
      name: 'foo',
      ensure: 'present',
      value: 'new value',
    },
  },
}
```

The `set` method updates resources to a new state. The `changes` parameter gets passed a hash of change requests, keyed by the resource's name. Each value is another hash with the optional `:is` and `:should` keys. At least one of the two has to be specified. The values will be of the same shape as those returned by `get`. After the `set`, all resources should be in the state defined by the `:should` values.

> Example for a `changes` request:
> ```
> changes = {
>   'name' => {
>     is: {
>       name: 'name',
>       ensure: 'present',

> >       value: 'original value,

Reviewer 3: missing  end quote

Author: :+1:

>     },
>     should: {
>       name: 'name',
>       ensure: 'present',

> >       value: 'new value,

Reviewer 3: missing  end quote

Author: :+1:

The `set` method should always return `nil`. Any progress signaling should be done through the logging utilities described below. If the `set` method throws an exception, all resources that should change in this call and haven't already been marked with a definite state, will be marked as failed. The runtime will only call the `set` method if there are changes to be made, especially in the case of resources marked with `noop => true` (either locally or through a global flag). The runtime will not pass them to `set`. See `supports_noop` below for changing this behaviour if required.

> Both methods take a `context` parameter which provides utilities from the runtime environment, and is described in more detail there.

Reviewer 2: Where is this described? In the runtime environment?

Author: yeah. changed it (and the three other occurrences) to "described in more detail in its own section below"

Once all of that is in place, instead of the `set` method, the provider needs to implement the `create`, `update` or `delete` methods:

* `create(context, name, should)`: This is called when a new resource should be created.

>   * `context`: provides provides utilities from the runtime environment, and is described in more detail there.

Reviewer: ```suggestion
  * `context`: provides utilities from the runtime environment, and is described in more detail there.
```

Author: :+1:

  * `name`: the name (if there is only one namevar) or hash of namevars (if there are multiple) of the new resource.
  * `should`: a hash of the attributes for the new instance.

* `update(context, name, should)`: This is called when a resource should be updated.

>   * `context`: provides provides utilities from the runtime environment, and is described in more detail there.

Reviewer: ```suggestion
  * `context`: provides utilities from the runtime environment, and is described in more detail there.
```

  * `name`: the name (if there is only one namevar) or hash of namevars (if there are multiple) of the new resource.
  * `should`: a hash of the desired state of the attributes.

* `delete(context, name)`: This is called when a resource should be deleted.

>   * `context`: provides provides utilities from the runtime environment, and is described in more detail there.

Reviewer: ```suggestion
  * `context`: provides utilities from the runtime environment, and is described in more detail there.
```

Author: the power of copy and paste :-1:

Once all of that is in place, instead of the `set` method, the provider needs to implement the `create`, `update` or `delete` methods:

* `create(context, name, should)`: This is called when a new resource should be created.
* `update(context, name, should)`: This is called when a resource should be updated.
* `delete(context, name)`: This is called when a resource should be deleted.

The parameters of these methods always carry the same values:

* `context`: provides utilities from the runtime environment, and is described in more detail in its own section below.

> * `name`: the name (if there is only one namevar) or hash of namevars (if there are multiple) of the new resource.

Reviewer 3: remove "new" at the end of this line. As should now read for update and delete

Author: :+1:

`SimpleProvider` requires that your type follows some common conventions:

> * the single namevar attribute is called `name` or the type has more than one namevar

Reviewer 3: This doesn't make sense to me as a sentence. Perhaps:

A single namevar attribute called `name`; or, `names` if the type has multiple namevar attributes, which will receive a hash.

Author: changed it back to "if the type has a single namevar, it must be called `name`"

looking at the code, I *think* we can relax that, too. but that's a conversation for another day.

Reviewer 3: 👍

The `canonicalize` method transforms its `resources` argument into the standard format required by the rest of the provider. The `resources` argument to `canonicalize` is an enumerable of resource hashes matching the structure returned by `get`. It returns all passed values in the same structure with the required transformations applied. It is free to reuse or recreate the data structures passed in as arguments. The runtime environment must use `canonicalize` before comparing user input values with values returned from `get`. The runtime environment always passes canonicalized values into `set`. If the runtime environment requires the original values for later processing, it protects itself from modifications to the objects passed into `canonicalize`, for example through creating a deep copy of the objects.

> The `context` parameter is the same passed to `get` and `set`, which provides utilities from the runtime environment, and is described in more detail there.

Reviewer: This could be rephrased for better readability. It is not `get` and `set` that "provides utilities"

Author: :+1:

# lib/puppet/provider/apt_key/apt_key.rb
class Puppet::Provider::AptKey::AptKey
  def get(context, names = nil)
    if names == nil
      return all_instances
    else
      names.collect { |n| find_instance(n) }
    end
  end
```

Some resources are very expensive to enumerate. The provider can implement `simple_get_filter` to signal extended capabilities of the `get` method to address this. The provider's `get` method will be called with an array of resource names, or `nil`. The `get` method must at least return the resources mentioned in the `names` array, but may return more than those. If the `names` parameter is `nil`, all existing resources should be returned. The `names` parameter defaults to `nil` to allow simple runtimes to ignore this feature.

> For types with multiple namevars, the `names` array will consist of hashes of the namevars and their variables instead of simple values:

Reviewer: This sounds odd. "namevars and their variables"...

Author: yup. should have been "namevars and their values". fixed.
